File: src/main/java/com/offz/spigot/custommobs/Pathfinders/Flying/PathfinderGoalIdleFlyAboveGround.java
package com.offz.spigot.custommobs.Pathfinders.Flying;

import com.offz.spigot.custommobs.Mobs.Flying.FlyingMob;

import java.util.Random;

public class PathfinderGoalIdleFlyAboveGround extends PathfinderGoalIdleFly {
    private double maxHeight;

    public PathfinderGoalIdleFlyAboveGround(FlyingMob flyingMob, int maxHeight) {
        super(flyingMob);
        this.maxHeight = maxHeight;
    }

    @Override
    public void c() {
        Random random = mob.getRandom();
        double dx = mob.locX + (double) ((random.nextFloat() * 2.0F - 1.0F) * 16.0F);
        double dy = mob.locY + (double) ((random.nextFloat() * 2.0F - 1.5F) * 16.0F); //make it more likely to fly down
        double dz = mob.locZ + (double) ((random.nextFloat() * 2.0F - 1.0F) * 16.0F);
        if (!mob.getBukkitEntity().getWorld().getBlockAt((int) dx, (int) (dy - maxHeight), (int) dz).getType().equals(org.bukkit.Material.AIR))
            mob.getControllerMove().a(dx, dy, dz, 1.0D);
    }
}

File: src/main/java/com/mineinabyss/mobzy/pathfinders/flying/PathfinderGoalIdleFlyAboveGround.kt
package com.mineinabyss.mobzy.pathfinders.flying

import com.mineinabyss.mobzy.mobs.types.FlyingMob
import org.bukkit.Location
import kotlin.random.Random

class PathfinderGoalIdleFlyAboveGround(mob: FlyingMob, private val maxHeight: Double = 4.0, private val radius: Double = 5.0) : PathfinderGoalIdleFly(mob) {

    override fun init() {
        val x = mob.locX + Random.nextDouble(-radius, radius)
        val y = mob.locY + Random.nextDouble(-radius / 4, radius) //make it more likely to fly down
        val z = mob.locZ + Random.nextDouble(-radius, radius)
        val loc = Location(entity.world, x, y, z)
        if (!loc.clone().add(0.0, -maxHeight, 0.0).block.type.isSolid) {
            moveController.a(mob.locX, mob.locY - 0.1, mob.locZ, 1.0)
            return
        }
        if (y > 16) { //keep mobs from going down and killing themselves
            targetLoc = loc
            moveController.a(x, y, z, 1.0) //TODO make a wrapper for the controller and figure out the difference between it and navigation
        }
    }
}